### PR TITLE
Load monster stats from csv file (Issue 41)

### PR DESCRIPTION
**Pull request by enEin100**
The new function LoadMonstatsFromFile will overwrite all the hard-coded values except for the names if the function UpdateStats fails.

### CODE REVIEW

**Comment by enEin100** [comment]
Any advice on where to look for how to pass the failed check would be appreciated, as well as feedback on my approach to issue 41.

File: src/fheroes2/monster/monster.cpp [hunk 1]
```diff
@@ -182,8 +183,58 @@ float Monster::GetUpgradeRatio(void)
...
+{
+    std::string line;
+    std::vector<std::string> stats;
+    std::string stat;
+    int stat_counter = 0;
+    std::ifstream config_file( "monster.csv" );
+
+    if ( config_file )
+	{
+        while ( config_file.good() )
```
Comment (ihhub) [inline src/fheroes2/monster/monster.cpp]: My suggestion is to create a temporary variable `std::vector< monstats_t > monsterStatistics` and push one by one `monstats_t` variables. After reading the file we need to verify that `monsterStatistics.size() == SIZE(monsters)`.

File: src/fheroes2/monster/monster.cpp [hunk 2]
```diff
@@ -182,8 +183,58 @@ float Monster::GetUpgradeRatio(void)
...
+		{
+            getline( config_file, line );
+
+			if ( line[0] == '/' )
+				continue;
+
+			else
+			{
+				std::stringstream cfg_line( line );
```
Comment (ihhub) [inline src/fheroes2/monster/monster.cpp]: We set counter to 0 here and after the loop we check the size of `stats` vector. If it's not equal 17 then it's bad we exit the function.

File: src/fheroes2/monster/monster.cpp [hunk 3]
```diff
@@ -182,8 +183,58 @@ float Monster::GetUpgradeRatio(void)
...
+                stat_counter += 17;
+			}
+
+			stat_counter = 0;
+		}
+
+        for ( int i = 0; i < stats.size() / 17; i++ )
+		{
+            monstats_t monster{
+			(u8)stoi( stats[2 + stat_counter] ), (u8)stoi( stats[3 + stat_counter] ),
```
Comment (ihhub) [inline src/fheroes2/monster/monster.cpp]: I don't think that this is a correct way to do type conversion. Please use `atoi` instead and then use C++-style type conversion (`static_cast<T>`) after checking retrieved value.

File: src/fheroes2/monster/monster.cpp [hunk 4]
```diff
@@ -182,8 +183,58 @@ float Monster::GetUpgradeRatio(void)
...
+
+            monsters[i] = monster;
+            stat_counter += 17;
+        }
+    }
+}
+
 void Monster::UpdateStats(const std::string & spec)
 {
+    bool xml_stats_loaded = true;
```
Comment (ihhub) [inline src/fheroes2/monster/monster.cpp]: Name it as `isXMLFileLoaded`

File: src/fheroes2/monster/monster.cpp [hunk 5]
```diff
@@ -222,10 +273,18 @@ void Monster::UpdateStats(const std::string & spec)
...
     }
     else
-    VERBOSE(spec << ": " << doc.ErrorDesc());
+	{
+		VERBOSE(spec << ": " << doc.ErrorDesc());
+        xml_stats_loaded = false;
+	}
 #endif
+
+	if ( xml_stats_loaded == false )
```
Comment (ihhub) [inline src/fheroes2/monster/monster.cpp]: `if ( !isXMLFileLoaded )`

File: src/fheroes2/monster/monster.h [hunk 6]
```diff
@@ -181,7 +181,10 @@ class Monster
 
     static u32 GetCountFromHitPoints(const Monster &, u32);
 
+	static void LoadMonstatsFromFile();
```
Comment (ihhub) [inline src/fheroes2/monster/monster.h]: Revert changes in this file. This function must not be accessible outside.

File: src/fheroes2/monster/monster.cpp [hunk 7]
```diff
@@ -182,8 +183,58 @@ float Monster::GetUpgradeRatio(void)
     return GameStatic::GetMonsterUpgradeRatio();
 }
 
+void Monster::LoadMonstatsFromFile()
```
Comment (ihhub) [inline src/fheroes2/monster/monster.cpp]: Put this function into unnamed namespace just after `monsters` variable declaration.

Comment (ihhub) [inline src/fheroes2/monster/monster.cpp]: Call this function as `LoadMonsterStatsFromCSVFile`

File: src/fheroes2/monster/monster.cpp [hunk 8]
```diff
@@ -182,8 +183,58 @@ float Monster::GetUpgradeRatio(void)
...
+{
+    std::string line;
+    std::vector<std::string> stats;
+    std::string stat;
+    int stat_counter = 0;
+    std::ifstream config_file( "monster.csv" );
```
Comment (ihhub) [inline src/fheroes2/monster/monster.cpp]: We have to pass an input parameter instead of hardcoding values.

**Comment by enEin100** [comment]
This helps a lot, thank you. I'll get right on it.

File: src/fheroes2/monster/monster.cpp (hunk 8 above)
Comment (enEin100) [reply src/fheroes2/monster/monster.cpp]: I'm not sure how to approach this one. Everything I looked up pointed to command line arguments and such.

File: src/fheroes2/monster/monster.cpp [hunk 9]
```diff
@@ -52,128 +53,183 @@ struct monstats_t
 namespace
 {
     monstats_t monsters[] = {
-	{     0,   0,   0,   0,   0,  Speed::VERYSLOW,   0,     0,    "Unknown Monster", "Unknown Monsters" , { 0, 0, 0, 0, 0, 0, 0} },
```
Comment (ihhub) [inline src/fheroes2/monster/monster.cpp]: Hi @enEin100 , could you please revert all changes related only for code formatting?

File: src/fheroes2/monster/monster.cpp (hunk 8 above)
Comment (enEin100) [reply src/fheroes2/monster/monster.cpp]: Figured out how to load my csv file like all other stats files and use a parameter for the file name

**Pull request closed without merging**